### PR TITLE
Pytorch Geometric integration

### PR DESCRIPTION
**Pull request by PabloAMC**
## What does this PR do?

PyTorch geometric integration, initially with graph classification support

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-lash/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/graph/GraphClassification/data.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+from flash.core.classification import ClassificationDataPipeline
+from flash.core.data.datamodule import DataModule
+from flash.core.data.utils import _contains_any_tensor
+
+'''
+The structure we follow is DataSet -> DataLoader -> DataModule -> DataPipeline
+'''
+
+class BasicGraphDataset(Dataset):
```
Comment (tchaton) [inline flash/graph/GraphClassification/data.py:33]: Not sure you need this one.

File: flash/graph/GraphClassification/data.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+    def len(self):
+        return len(self.processed_file_names)
+
+    def get(self, idx):
+        data = torch.load(os.path.join(self.processed_dir, 'data_{}.pt'.format(idx)))
+        #TODO: Is data.pt the best way/file type to load the data? 
+        #TODO: Interface with networkx would probably go here with some option to say how to load it
+        return data
+class FilepathDataset(torch.utils.data.Dataset):
```
Comment (tchaton) [inline flash/graph/GraphClassification/data.py:72]: I like this idea :)

File: flash/graph/GraphClassification/data.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+        datamodule = cls(
+            train_ds=train_ds,
+            valid_ds=valid_ds,
+            test_ds=test_ds,
+            batch_size=batch_size,
+            num_workers=num_workers,
+        )
+
+        datamodule.num_classes = len(train_ds.classes)
+        datamodule.data_pipeline = GraphClassificationDataPipeline(
```
Comment (tchaton) [inline flash/graph/GraphClassification/data.py:377]: Where is it defined ?

File: flash/graph/GraphClassification/model.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        learning_rate: float = 1e-3,
+    ):
+
+        if isinstance(hidden, int):
+            hidden = [hidden]
+
+        #sizes = [input_size] + hidden + [num_classes]
+
+        super().__init__(
+            model = GCN(in_features = num_features, hidden_channels=hidden, out_features = num_classes),
```
Comment (tchaton) [inline flash/graph/GraphClassification/model.py:58]: Users should be able to provide their own Graph model.

File: flash/graph/GraphClassification/model.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        )
+
+    #train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
+
+    def forward(self, data) -> Any:
+        x = self.model(data.x, data.edge_index, data.batch) #This line is probably something to change
+        return self.head(x)
+
+    @staticmethod
+    def default_pipeline() -> ClassificationDataPipeline:
```
Comment (tchaton) [inline flash/graph/GraphClassification/model.py:72]: You need a custom `GraphClassificationDataPipeline` to be able to load from different time of raw graph data. Just need to override the before_collate function.

**Review by tchaton — COMMENTED** [review]
Awesome work ! Almost there !

**Comment by aribornstein** [comment]
Whats the status on this?

File: flash/graph/GraphClassification/data.py (hunk 3 above)
Comment (PabloAMC) [reply flash/graph/GraphClassification/data.py:377]: I have included it, although will most likely need to be changed since it is mostly the ImageClassificationPipeline

File: flash/graph/GraphClassification/data.py (hunk 1 above)
Comment (PabloAMC) [reply flash/graph/GraphClassification/data.py:33]: Yeah, I'm not sure either. This is the way one would build a DataSet using the prescription from Pytorch Geometric. But perhaps it is not needed

**Comment by Borda** [comment]
@PabloAMC @tchaton how is it going here?

File: flash_examples/finetuning/graph_classification.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+from flash.graph.classification.data import GraphClassificationData
+from flash.graph.classification.model import GraphClassifier
+
+if _PYTORCH_GEOMETRIC_AVAILABLE:
+    import torch_geometric.transforms as T
+    from torch_geometric.datasets import TUDataset
+else:
+    raise ModuleNotFoundError("Please, pip install -e '.[graph]'")
+
+dataset = TUDataset("data", name='IMDB-BINARY')
```
Comment (rusty1s) [inline flash_examples/finetuning/graph_classification.py:14]: It might be a good idea to shuffle the dataset before creating train/test splits, e.g.:
```python
dataset = TUDataset("data", name='IMDB-BINARY').shuffle()
```

File: requirements/datatype_graph.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+torch-scatter
+torch-sparse
+torch-cluster
```
Comment (rusty1s) [inline requirements/datatype_graph.txt:3]: `torch-cluster` is not required for this example.

**Review by rusty1s — CHANGES_REQUESTED** [review]
This is super cool, thanks a lot for this effort :) I only have minor requests.

**Edit:** Wondering whether it's really necessary to push the dataset to the repository, as the data will be automatically downloaded.

**Comment by PabloAMC** [comment]
@ethanwharris I think this branch is ready for review. The only thing I am not fully sure about is the `GraphSequenceDataSource()` and class method `from_pygdatasequence`  in DataModule; and the corresponding tests. The rest was mostly done by @tchaton  :)

File: flash/graph/classification/data.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    def get_state_dict(self) -> Dict[str, Any]:
+        return self.transforms
+
+    @classmethod
+    def load_state_dict(cls, state_dict: Dict[str, Any], strict: bool = False):
+        return cls(**state_dict)
+
+    @staticmethod
+    def default_transforms() -> Optional[Dict[str, Callable]]:
+        return {"pre_tensor_transform": NormalizeFeatures(), "collate": Batch.from_data_list}
```
Comment (tchaton) [inline flash/graph/classification/data.py:57]: Neat !

File: flash_examples/graph_classification.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+if _TORCH_GEOMETRIC_AVAILABLE:
+    from torch_geometric.datasets import TUDataset
+else:
+    raise ModuleNotFoundError("Please, pip install -e '.[graph]'")
+
+# 1. Create the DataModule
+dataset = TUDataset(root="data", name="KKI")
+
+datamodule = GraphClassificationData.from_datasets(
```
Comment (tchaton) [inline flash_examples/graph_classification.py:27]: Would be nice to add `from_folders` support.

**Review by tchaton — APPROVED** [review]
Looks great ! Awesome work.

File: flash_examples/graph_classification.py (hunk 9 above)
Comment (PabloAMC) [reply flash_examples/graph_classification.py:27]: I created a branch on my repository for the `from_sequence` support, so perhaps I can add it there and create a PR :-)

File: flash/core/data/auto_dataset.py [hunk 10]
```diff
@@ -93,7 +93,7 @@ class AutoDataset(BaseAutoDataset[Sequence], Dataset):
     must be a ``Sequence`` (it must have a length)."""
 
     def __getitem__(self, index: int) -> Any:
-        return self._call_load_sample(self.data[index])
+        return self._call_load_sample(self.data[int(index)])
```
Comment (ethanwharris) [inline flash/core/data/auto_dataset.py:96]:
```suggestion
        return self._call_load_sample(self.data[index])
```

**Review by ethanwharris — APPROVED** [review]
